RedfishPkg/RedfishCredentialDxe/RedfishCredentialDxe.c

```diff
@@ -123,6 +130,664 @@ RedfishCredentialStopService (
+
+  if (mCredentialPrivate == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: failed with error - %r\n", __func__, EFI_NOT_STARTED));
+    return EFI_NOT_STARTED;
+  }
+
+  if ((ServiceStopType == ServiceStopTypeExitBootService) ||
+      (ServiceStopType == ServiceStopTypeNone))
+  {
+    // Check PCD and skip the action if platform library is responsible for deleting account
+    // on exit boot service event
+    if (FixedPcdGetBool (PcdRedfishCredentialDeleteAccount)) {
+      if (!IsListEmpty (&mCredentialPrivate->RedfishServiceList)) {
+        Instance = (REDFISH_SERVICE_LIST *)GetFirstNode (&mCredentialPrivate->RedfishServiceList);
+        IterateThroughBootstrapAccounts (Instance->RedfishService);
+      }
+
+      ReleaseCredentialPrivate ();
+    }
+  }
+
+  Status = mCredentialPrivate->RedfishCredentialProtocol.StopService (
+                                                           &mCredentialPrivate->RedfishCredentialProtocol,
+                                                           ServiceStopType
+                                                           );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to stop service - %r\n", __func__, Status));
+    return Status;
+  }
+
+  return Status;
+}
+
+/**
+  Function sends DELETE request to BMC for the account defined by the target URI.
+
+  @param[in]  RedfishService       Pointer to Redfish Service to be used
+                                   for sending DELETE request to BMC.
+  @param[in]  TargetUri            URI of bootstrap account to send DELETE request to.
+
+**/
+EFI_STATUS
+EFIAPI
+DeleteRedfishBootstrapAccount (
+  IN REDFISH_SERVICE  RedfishService,
+  IN CHAR16           *TargetUri
+  )
+{
+  EFI_STATUS        Status;
+  REDFISH_RESPONSE  RedfishResponse;
+
+  if (mCredentialPrivate == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: failed with error - %r\n", __func__, EFI_NOT_STARTED));
+    return EFI_NOT_STARTED;
+  }
+
+  if ((RedfishService == NULL) || (mCredentialPrivate->AuthMethod != AuthMethodHttpBasic)) {
+    DEBUG ((DEBUG_ERROR, "%a: Redfish service is not available\n", __func__));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Remove bootstrap account at /redfish/v1/AccountService/AccountId
+  //
+  ZeroMem (&RedfishResponse, sizeof (REDFISH_RESPONSE));
+  Status = RedfishHttpDeleteResourceEx (
+             RedfishService,
+             TargetUri,
+             "{}",
+             2,
+             NULL,
+             &RedfishResponse
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: can not remove bootstrap account at BMC: %r", __func__, Status));
+    DumpRedfishResponse (__func__, DEBUG_ERROR, &RedfishResponse);
+  } else {
+    DEBUG (
+      (REDFISH_CREDENTIAL_DEBUG, "%a: bootstrap account: %a is removed from: %s\nURI - %s",
+       __func__, mCredentialPrivate->AccountName, REDFISH_MANAGER_ACCOUNT_COLLECTION_URI, TargetUri)
+      );
+  }
+
+  RedfishHttpFreeResponse (&RedfishResponse);
+

```

Comment:
Can we explicitly close HTTP connection here because this connect is no longer usable? This helps to end the connection immediately instead of waiting for timeout at BMC side. Below is my example of how to close connection:

https://github.com/tianocore/edk2-redfish-client/blob/952064cc1bba5a066652d1fa3d53f475e7f69078/RedfishClientPkg/RedfishBootstrapAccountDxe/RedfishBootstrapAccountDxe.c#L154-L161

Reply:
In you implementation you have Redfish Config Handler Protocol implemented by your driver. So, you got that Redfish Service pointer used exclusively by your driver.
In Redfish Credential driver we have that Redfish Service created and registered by external driver. In our case it is Redfish HTTP driver. What if we close connection in Redfish HTTP driver in RedfishFreeRedfishService function? Right after we call Unregister function of Redfish Credential Protocol?

Reply:
Closing connection in Redfish HTTP driver sounds good to me. But we might want to handle the case when RedfishFreeRedfishService () is called multiple times by feature drivers and we need to close connection only once (since the RestEx is shared for all feature drivers)

Reply:
Hi Nickle,
I added a closing connection into RedfishFreeRedfishService() function and I close connection if UnregisterRedfishService function finished with success. So, if we call RedfishFreeRedfishService with the RedfishService parameter which was closed before, UnregisterRedfishService will return No Found error and I will not close connection.